### PR TITLE
2021 10 01 fix null/missing values on percentage tooltip

### LINKED ISSUE #1414: Stacked as percentage reports null/missing values as 0
**Describe the bug**

**To Reproduce**
1. Go to https://codesandbox.io/s/eloquent-wildflower-lovci?file=/src/App.tsx
2. Mouse over the column `b`,`c` the tooltip should not show the `null` b value and should not show the missing `c` value for the `b` series.
3. Remove the `stackMode` and verify that the `b` and `c` bars tooltip correctly hides the missing/null values.

**Expected behaviour**

In both, percentage mode and stacked mode, the tooltip should reflect the actual values:
- if the value is a valid number, it should be shown
- if the value is `null` or `undefined` no value should be shown

**Screenshots**

normally stacked 

percentage stacked

**Version (please complete the following information):**
 - Elastic Charts: > 33.0.0

**Additional context**

**Errors in browser console**
If applicable, add any console error.

### PR DESCRIPTION
**Pull request by markov00**
## Summary

A regression was introduced and caused null values to show up on the tooltip when rendering stacked bars and areas.
We fixed that issue including proper testing for the use case.

## Details
The file `chart_types/xy_chart/state/selectors/get_tooltip_values_highlighted_geoms.ts` is used to pick the tooltip values from the subset of geometries that the current mouse position intersect. A check was introduced in #1226 to check whenever to include or not a null Y value into a tooltip list.
That PR also altered the current data flow, allowing nullish value in the `y1` or `initialY1` to flow down within the `indexedGeometryMap` to be used on the tooltip, but filtered on the rendering side.
This together with a wrong logic and no test on that edge case caused the regression.

## Issues

fix #1414 

### Checklist

- [x] The proper **chart type** label has been added (e.g. `:xy`, `:partition`)
- [x] The proper **feature** labels have been added (e.g. `:interactions`, `:axis`)
- [x] All related issues have been linked (i.e. `closes #123`, `fixes #123`)
- [x] Unit tests have been added or updated to match the most common scenarios
- [x] The proper documentation and/or storybook story has been added or updated

### CODE REVIEW

File: packages/charts/src/chart_types/xy_chart/rendering/bars.ts [hunk 1]
```diff
@@ -87,8 +88,8 @@ export function renderBars(
       const width = clamp(seriesStyle.rect.widthPixel ?? xScale.bandwidth, minPixelWidth, maxPixelWidth);
       const x = xScaled + xScale.bandwidth * orderIndex + xScale.bandwidth / 2 - width / 2;
 
-      const originalY1Value = stackMode === StackMode.Percentage ? (isNil(y1) ? null : y1 - (y0 ?? 0)) : initialY1;
-      const formattedDisplayValue = displayValueSettings?.valueFormatter?.(originalY1Value);
+      const y1Value = getDatumYValue(datum, false, false, stackMode);
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/xy_chart/rendering/bars.ts:91]: Is the 3rd arg (`isBandChart`) always false here?

**Review by nickofthyme — APPROVED** [review]
LGTM, tested locally.

File: packages/charts/src/chart_types/xy_chart/rendering/bars.ts (hunk 1 above)
Comment (markov00) [reply packages/charts/src/chart_types/xy_chart/rendering/bars.ts:91]: Bar chart can be also a band chart, but we only show a single value on the legend/tooltip that reflect at the moment the `initialY1` or the calculated percentage.
Let me check if I can quickly fix that providing: two values for a single band bar (upper/lower)

File: storybook/stories/bar/12_stacked_as_percentage.story.tsx [hunk 2]
```diff
@@ -14,17 +14,17 @@ import { Axis, BarSeries, Chart, Position, ScaleType, Settings, StackMode } from
 import { useBaseTheme } from '../../use_base_theme';
 
 export const Example = () => {
-  const stackedAsPercentage = boolean('stacked as percentage', true);
-  const clusterBars = boolean('cluster', true);
+  const stackMode = boolean('stacked as percentage', true) ? StackMode.Percentage : undefined;
```
Comment (monfera) [inline storybook/stories/bar/12_stacked_as_percentage.story.tsx:17]: Unrelated: what's the meaning of `StackMode` in cases where it'd otherwise apply, but its value is `undefined`?

Comment (markov00) [reply storybook/stories/bar/12_stacked_as_percentage.story.tsx:17]: Its a mix up of various things:
- the `undefined` one is a simple stack where values are summed
- the `StackMode.percentage` computes and use the ratio of the values within each bucket separately
- `StackMode.Silhouette and Wiggle` represents two different stack offsets (see d3 stack offset https://github.com/d3/d3-shape#stack-offsets used by visualize and useful for stream graphs)

File: packages/charts/src/chart_types/xy_chart/rendering/points.ts [hunk 3]
```diff
@@ -160,30 +160,25 @@ export function getPointStyleOverrides(
...
+    // that are already computed and available on y0 and y1
+    // in all other cases for band charts, I want to get back the original/initial value of y0 and y1
+    // not the computed value
+    return stackMode === StackMode.Percentage ? (lookingForY0 ? y0 : y1) : lookingForY0 ? initialY0 : initialY1;
   }
   // if not a band chart get use the original/initial value in every case except for stack as percentage
   // in this case, we should take the difference between the bottom position of the bar and the top position
   // of the bar
-  return stackMode === StackMode.Percentage ? (y1 ?? 0) - (y0 ?? 0) : initialY1;
+  return stackMode === StackMode.Percentage ? (isNil(y1) || isNil(initialY1) ? null : y1 - (y0 ?? 0)) : initialY1;
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/chart_types/xy_chart/rendering/points.ts:181]: let me pull master now that your PR is merged and refresh the code
